src/database_observer.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Database agent relation observer."""
+import logging
+import typing
+
+import ops
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseCreatedEvent,
+    DatabaseEndpointsChangedEvent,
+    DatabaseRequires,
+)
+from ops.charm import CharmBase
+from ops.framework import Object
+
+from charm_types import DatasourcePostgreSQL
+from constants import SYNAPSE_CONTAINER_NAME
+from database_client import DatabaseClient
+from exceptions import CharmDatabaseRelationNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseObserver(Object):
+    """The Database relation observer.
+
+    Attrs:
+        _pebble_service: instance of pebble service.
+    """
+
+    _RELATION_NAME = "database"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        # SUPERUSER is required to update pg_database
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+            database_name=self._charm.app.name,
+            extra_user_roles="SUPERUSER",
+        )
+        self.framework.observe(self.database.on.database_created, self._on_database_created)
+        self.framework.observe(self.database.on.endpoints_changed, self._on_endpoints_changed)
+
+    @property
+    def _pebble_service(self) -> typing.Any:
+        """Return instance of pebble service.
+
+        Returns:
+            instance of pebble service or none.
+        """
+        return getattr(self._charm, "pebble_service", None)
```

Comment:
This doesn't quite feel right yet, it would be better not to need this. I believe this comes from that pebble service requires the charm state, charm state requires information from the database and the database observer requires the pebble service.

From memory, one of the options was to instantiate the `DatabaseRequires` in `charm.py` and pass that to `CharmState` for it to retrieve the necessary data from it if the relation is defined and also pass `DatabaseRequires` as an argument to the initialiser of this class. That way `CharmState` doesn't depend on the database observer anymore and it is safe to pass `CharmState` as an argument to the database observer which then means that the pebble service doesn't need to keep track of charm state and can be passed as an argument to the database observer as well

Reply:
Actually, this is here because after getting information from the relation with Database, the configuration must be updated and the container replaned.
This way avoids a circular dependency.
The idea was gently offered by Yanks at the time :-) : https://github.com/canonical/jenkins-agent-k8s-operator/blob/d094425a33a30b0bd08d36849d949c555e2f1511/src/agent.py#L20

Reply:
Related: #54